src/MaxText/layers/moe.py

```diff
@@ -579,10 +579,13 @@ def unpermute(
       if self.config.decoder_block == ctypes.DecoderBlockType.LLAMA4:
         # For Llama4, combine using weights of 1 for selected experts
         reshaped_weights = jnp.ones_like(reshaped_weights)
+      if self.config.weight_sum_fp32:
+        reshaped_intermediate = reshaped_intermediate.astype(jnp.float32)
+        reshaped_weights = reshaped_weights.astype(jnp.float32)
       output = jnp.einsum(
           "BKE,BK -> BE",
-          reshaped_intermediate.astype(jnp.float32),
-          reshaped_weights.astype(jnp.float32),
+          reshaped_intermediate,
```

Comment:
Could you also help update this block at line #1616?

```
      with jax.named_scope("w_sum"):
        if is_llama4_decoder_layer:
          weights = self.reshape_and_update_weights(jnp.ones_like(top_k_weights), top_k_indices)
        # cast to f32 for sum up in einsum op
        output = jnp.einsum(
            "BSEM,BSE -> BSM",
            intermediate_layer.astype(jnp.float32),
            weights.astype(jnp.float32),  # pylint: disable=undefined-variable,possibly-used-before-assignment
            precision=matmul_precision,
        ).astype(self.dtype)
```

Reply:
Thanks, just added that part as well